### ISSUE TITLE
Aarch64 macOS (Apple Silicon) support

### ISSUE THREAD

**Issue by mcginty** [exelix11/SysDVR#166]
I'm running macOS 12.2, and I've installed libusb and dotnet via `brew install --cask dotnet` and `brew install sdl2 libusb libusb-compat`.
Running `SysDVR-Client.dll usb` leads to an error about `LibUsbDotNet` being missing.

**What's the error message you get ?**
```
$ dotnet --info

Host (useful for support):
  Version: 6.0.1
  Commit:  3a25a7f1cc

.NET SDKs installed:
  No SDKs were found.

.NET runtimes installed:
  Microsoft.NETCore.App 6.0.1 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]

$ DYLD_LIBRARY_PATH=/opt/homebrew/lib dotnet SysDVR-Client.dll usb
SysDVR-Client - 5.2 by exelix
https://github.com/exelix11/SysDVR

Initializing video player with h264 codec.
System.IO.FileNotFoundException: Could not load file or assembly 'LibUsbDotNet, Version=1.0.0.0, Culture=neutral, PublicKeyToken=c677239abe1e02a9'. The system cannot find the file specified.

File name: 'LibUsbDotNet, Version=1.0.0.0, Culture=neutral, PublicKeyToken=c677239abe1e02a9'
   at SysDVR.Client.Program.OpenUsbSource(LogLevel usbLogLeve, Boolean forceLibUsb, String preferredSerial)
   at SysDVR.Client.Program.InnerMain(String[] args)
   at SysDVR.Client.Program.Main(String[] args)
```

Note that `SysDVR-Client.dll bridge` seems to work and opens a blank window as expected.

**Comment by exelix11** [comment]
Ok i've been looking into this, it seems libusbdotnet which is the library i use for USB communications is not building for arm osx.
The problem is that they target an older dotnet standard which doesn't support conditional compilation for arm on osx which breaks including the libusb dylib.

The workaround is to use dotnet x64 over rosetta so it loads the currently included x64 version of libusbdotnet.

As i don't even have a mac to test idk if i'll be able to fix this, will consider opening an issue on the libusbdotnet side.

**Comment by mcginty** [comment]
Thanks for looking into it! That makes sense. I did manage to get it working using Rosetta but ended up on another `libusb: error [submit_bulk_transfer] bulk transfer failed (dir = In): pipe is stalled (code = 0xe000404f)` error :P.

No worries and thanks for working on such a cool project!

**Comment by janerjak** [comment]
@mcginty Don't know if this is still relevant, but I got it working on my M1 Pro. (No audio output though. There is audio being transmitted, but it is only briefly audible, when the stream window is closed).

What issues I had and what helped:
- .NET 6 Runtime just doesn't work for me. Always ends up crashing / does not find dependencies that are clearly at the correct path. So I installed .NET 5 and that works fine (x86 runtime or SDK).
- As described [here](https://github-wiki-see.page/m/exelix11/SysDVR/wiki/SysDVR-Client-setup), I installed the dependencies using brew (make sure to use your x86 brew installation). However, the latest ffmpeg-build in brew installs `libavcodec59.dylib` instead of the used `libavcodec58.dylib`. After checking around in version histories for ffmpeg, I finally found that `arch -x86_64 /path/to/intel/brew uninstall ffmpeg; arch -x86_64 /path/to/intel/brew install ffmpeg@4;` installs the correct shared library, but in keg-only mode. Make sure to use `arch -x86_64 /path/to/intel/brew link ffmpeg@4` afterwords, to create the necessary symlinks.
- It always failed at "libusb-1.0.0.dylib cannot be opened because the developer cannot be verified", so I had to install libusb (x86) through brew as well and copy the shared library file in the Cellar-folder to the SysDVR folder manually.
- And finally, make sure to use the correct dotnet version. I found no easy way to uninstall .NET 6 (the installer package does not recognize an already present install), so I just deleted all folders associated with it and then installed .NET 5 again. Verify by using `dotnet --version`.

(Also for some reason, the Switch .nro is really bugged for me and I have to reboot after a change or else it says to already be switching the mode, but it is not. Launching a game does not resolve this, only a hard-reboot).

If you still have issues, maybe we can help. Good luck :)

**Comment by KazushiMe** [comment]
Can reproduce on macOS aarch64. Have tried other solutions but none would work:

1. Windows on ARM VM via Parallels Desktop: zadig driver is only available for x86(-64), cannot be installed on arm64
2. Ubuntu Linux (ARM64) VM via Parallels Desktop: USB device is detected in `lsusb`, but `dotnet SysDVR-Client.dll usb` shows black screen for usb mode (tcp bridge works though)
3. Asahi Linux: No OpenGL acceleration, black screen whatever mode is selected.

I'm wondering if the client (C#) could be implemented in other languages that macOS natively supports (preferably in Swift).

**Comment by exelix11** [comment]
> Windows on ARM VM via Parallels Desktop: zadig driver is only available for x86(-64), cannot be installed on arm64

Windows on arm is not really a big thing so that's to be expected 

> Asahi Linux: No OpenGL acceleration, black screen whatever mode is selected.
I think you can force SDL to run without acceleration with environment variables, now i don't remember if i have that as a CLI option too.

The real problem is that the client currently uses .net 5 which did not support M1 macs. I'm planning to upgrade to .net 6 which brings full compatibility but i don't have a device to test on. 
So i'll ask here too, if anyone wants to help me test on M1 DM me on discord and i'll start working on porting it.

**Comment by janerjak** [comment]
Do you mean porting it to run natively on .NET 6 ARM? SysDVR currently works using Rosetta 2.

**Comment by exelix11** [comment]
Yeah going for native .net 6.
The problem is not really on sysdvr side but on the libs i use, i'd have to repackage libusb sdl and ffmpeg (which i already forked for arm32 support on linux and other stuff) to support .net 6 and the mac aarc64 platform then it should hopefully just work.

I've asked for someone to test on m1 several times but nobody replied so i just stashed this idea.

**Comment by janerjak** [comment]
> Can reproduce on macOS aarch64. Have tried other solutions but none would work:
> 
> 1. Windows on ARM VM via Parallels Desktop: zadig driver is only available for x86(-64), cannot be installed on arm64
> 2. Ubuntu Linux (ARM64) VM via Parallels Desktop: USB device is detected in `lsusb`, but `dotnet SysDVR-Client.dll usb` shows black screen for usb mode (tcp bridge works though)
> 3. Asahi Linux: No OpenGL acceleration, black screen whatever mode is selected.
> 
> I'm wondering if the client (C#) could be implemented in other languages that macOS natively supports (preferably in Swift).

Did you try my suggested solution?

**Comment by janerjak** [comment]
> I've asked for someone to test on m1 several times but nobody replied so i just stashed this idea.

I'd be glad to help in that regard. Just messaged you on Discord, but it requires you to accept the friends request. Darkn#*

**Comment by KazushiMe** [comment]
> Did you try my [suggested solution](https://github.com/exelix11/SysDVR/issues/166#issuecomment-1053022319)?

Forgot to mention I'm using dotnet v5 x86-64 runtime on macOS 11.5.2 and facing no issues other than libusb one (`libusb: error [submit_bulk_transfer] bulk transfer failed (dir = In): pipe is stalled (code = 0xe000404f)`).

> So i'll ask here too, if anyone wants to help me test on M1 DM me on discord and i'll start working on porting it.

DMed and willing to help.

**Comment by exelix11** [comment]
@KazushiMe and @janerjak can you try [this build](https://github.com/exelix11/SysDVR/files/8508397/net6.0.zip) ? I just built for .net 6 and upgraded to latest libusb which already includes mac arm binaries.
FYI this won't work on windows cause i removed the ffmpeg dlls to save space.

**Comment by KazushiMe** [comment]
> @KazushiMe and @janerjak can you try [this build](https://github.com/exelix11/SysDVR/files/8508397/net6.0.zip) ? I just built for .net 6 and upgraded to latest libusb which already includes mac arm binaries.

It works on native dotnet 6 with sdl2@2.0.20 and ffmpeg@4.4.1.

Have to set env variable to resolve dylibs though: `export DYLD_FALLBACK_LIBRARY_PATH=/opt/homebrew/Cellar/sdl2/2.0.20/lib:/opt/homebrew/Cellar/ffmpeg@4/4.4.1/lib`

**Comment by exelix11** [comment]
Ok the fact that it works at all is already great. 
No clue about stuttering, the usb3 thing could be that on sysdvr side i'm not configuring something properly and mac os gets mad about it, it's not something i test with but i did hear that it works on Windows.

Will start by adding the upgraded components to the repo then i'll decide what to do.

**Comment by KazushiMe** [comment]
> And I do notice some regular stutter with usb mode (does not happen with tcp bridge, and there is no stutter in the video saved to album).

Here's the screen recording and `--usb-debug` log. Stutter started at [97.31s], resumed at [97.97s]
[usb-debug.log.zip](https://github.com/exelix11/SysDVR/files/8518370/usb-debug.log.zip)
https://user-images.githubusercontent.com/85604869/164153592-23a54966-a058-4bff-9b8d-b288238f5440.mp4

Seems no libusb error occurred. Have tried `--file` to export mp4, only usb mode would produce such stutters (tcp bridge is fine). So at least it is not due to client video playback or sysmodule.

**Comment by exelix11** [comment]
Can you try [this](https://github.com/exelix11/SysDVR/files/8520513/net6.0.zip) just launch in usb mode with no other args, it should log the timings for each USB operation.

**Comment by KazushiMe** [comment]
> Can you try [this build](https://github.com/exelix11/SysDVR/files/8520513/net6.0.zip) just launch in usb mode with no other args, it should log the timings for each USB operation.

```
[Video] [ReadHeader] SendReq 0 ms | RecBody 16 ms
Warning: libusb error Timeout while reading header
[Video] [ReadHeader] SendReq 0 ms | RecBody 664 ms
[Video] [ReadHeader] SendReq 0 ms | RecBody 7 ms
```

**Comment by exelix11** [comment]
Well i'm not sure how to fix this, it just seems to fail to receive some data (that or the request doesn't get through). How often does this happen ?

**Comment by KazushiMe** [comment]
> How often does this happen ?

About 6 times per minute and it's quite annoying.
I have tried on a Windows x86-64 machine, and it does not stutter that much - at most twice per min and not that perceivable (no more than 100ms).

**Comment by exelix11** [comment]
I reworked the USB protocol to do less communication [client.zip](https://github.com/exelix11/SysDVR/files/8531739/client.zip) [Sysmodule.zip](https://github.com/exelix11/SysDVR/files/8531741/Sysmodule.zip). Update both client and sysmodule and let me know if it works.
I think this is the best i can do, if there's still stuttering it's probably some libusb problem i don't have control over.

**Comment by exelix11** [comment]
Did background playback work before ? I don't see what could have broken

**Comment by KazushiMe** [comment]
> Did background playback work before ? I don't see what could have broken

Nope, it didn't happen before. It's quite weird and has different trigger conditions on Big Sur and Monterey. I'd say it is more likely the fault of darwin kernel.

Also when SIP is disabled, macOS does not allow embeded `libusb-1.0.dylib` to load, requiring developer to update. This one built via Xcode works though: [libusb-1.0.dylib.osx-arm64.zip](https://github.com/exelix11/SysDVR/files/8532713/libusb-1.0.dylib.osx-arm64.zip)

**Comment by exelix11** [comment]
Yeah I'm probably removing that libusb and updating the guide to make you install it with brew, that should so e verification issues.

About the background issue will try some stuff, can you attach a longer log?

**Comment by KazushiMe** [comment]
> About the background issue will try some stuff, can you attach a longer log?

[log.txt.zip](https://github.com/exelix11/SysDVR/files/8533212/log.txt.zip)
Started at `[ 3.027695]`.
And I found the trigger condition: client SDL2 window is fully covered by another window.

**Comment by exelix11** [comment]
Added some more time before timing out and hopefully better handling of disconnection so it should resume [SysDVR-Client.zip](https://github.com/exelix11/SysDVR/files/8534150/SysDVR-Client.zip) (this is just the code dll file, replace the one with the same name in the build you arlready have)
Was thinking, this laggy behavior could also be caused by osx reducing the client priority if it's not in foreground, can you try manually increasing its priority from the task manager-equivalent and see what happens ? Just the first step above normal should do.

**Comment by KazushiMe** [comment]
This build does resume in 3 sec when changing state from foreground to background without introducing any stutter.

Have tried setting niceness to lowest "-20", QoS to "User-interactive" (`taskpolicy -b -p PID`) or disabling AppNap, but without success.

CPU Profiler logging via Instruments:

<img width="1526" alt="Instruments" src="https://user-images.githubusercontent.com/85604869/164572785-84ac874a-5a46-40c5-8f26-72722770b33c.png">

I suppose this behavior is expected unless developer specifies priority or taskpolicy explicitly in Obj-C or Swift.

```
➜ sudo taskinfo 4245
process: "dotnet" [4245] [unique ID: 4245]
architecture: arm64
coalition (type 0) ID: 675
coalition (type 1) ID: 676
suspend count: 0
virtual bytes: 393.13 GB; phys_footprint bytes: 127.46 MB; phys_footprint lifetime maximum bytes: 157.63 MB
run time: 769 s
user/system time    (current threads): 238.303259 s / 47.628466 s
user/system time (terminated threads): 0.034207 s / 0.001925 s
interrupt wakeups: 210315 (488 / 0.23% from platform idle)
default sched policy: POLICY_TIMESHARE
CPU usage monitor: none
CPU wakes monitor: 150 wakes per second (over system-default time period)
dirty tracking: untracked  dirty
boosts: 1 (1 externalized)
requested policy
	req apptype: TASK_APPTYPE_APP_DEFAULT
	req role: TASK_BACKGROUND_APPLICATION (PRIO_DARWIN_ROLE_UI_NON_FOCAL)
	req qos clamp: THREAD_QOS_UNSPECIFIED
	req base/override latency qos: LATENCY_QOS_TIER_0 / LATENCY_QOS_TIER_UNSPECIFIED
	req base/override thruput qos: THROUGHPUT_QOS_TIER_0 / THROUGHPUT_QOS_TIER_UNSPECIFIED
	req darwin BG: NO  
	req internal/external iotier: THROTTLE_LEVEL_TIER0 (IMPORTANT) / THROTTLE_LEVEL_TIER0 (IMPORTANT)
	req darwin BG iotier: THROTTLE_LEVEL_TIER2 (UTILITY)
	req managed: NO
	req other: boosted 
	req suppression (App Nap) behaviors: timer-throttling (LATENCY_QOS_TIER_0) 
effective policy
	eff role: TASK_BACKGROUND_APPLICATION (PRIO_DARWIN_ROLE_UI_NON_FOCAL)
	eff latency qos: LATENCY_QOS_TIER_0
	eff thruput qos: THROUGHPUT_QOS_TIER_0
	eff darwin BG: NO
	eff iotier: THROTTLE_LEVEL_TIER0 (IMPORTANT)
	eff managed: NO
	eff qos ceiling: THREAD_QOS_UNSPECIFIED
	eff qos clamp: THREAD_QOS_UNSPECIFIED
	eff other: live-donor 
imp_donor: CURRENTLY
imp_receiver: DE-NAP
pid suspended: NO
adaptive daemon: NO (boosted: NO)
```

**Comment by exelix11** [comment]
Set the priority of the thread that deals with USB to the max value allowed by .net, hopefully it helps [SysDVR-Client.zip](https://github.com/exelix11/SysDVR/files/8543219/SysDVR-Client.zip)

**Comment by KazushiMe** [comment]
> Set the priority of the thread that deals with USB to the max value allowed by .net, hopefully it helps [SysDVR-Client.zip](https://github.com/exelix11/SysDVR/files/8543219/SysDVR-Client.zip)

Still the same, but currently it should cover normal usage unless user alt+tab frequently. 

I'm leaving other info here for further investigation:

CLI interface only `taskinfo`:
```
effective policy
	eff role: TASK_UNSPECIFIED (PRIO_DARWIN_ROLE_DEFAULT)
	eff latency qos: LATENCY_QOS_TIER_UNSPECIFIED
	eff thruput qos: THROUGHPUT_QOS_TIER_UNSPECIFIED
	eff darwin BG: NO
	eff iotier: THROTTLE_LEVEL_TIER0 (IMPORTANT)
	eff managed: NO
	eff qos ceiling: THREAD_QOS_LEGACY
	eff qos clamp: THREAD_QOS_UNSPECIFIED
	eff other: live-donor 
```
With SDL2 window created `taskinfo`:
```
effective policy
	eff role: TASK_FOREGROUND_APPLICATION (PRIO_DARWIN_ROLE_UI_FOCAL) -> TASK_BACKGROUND_APPLICATION (PRIO_DARWIN_ROLE_UI_NON_FOCAL)
	eff latency qos: LATENCY_QOS_TIER_0
	eff thruput qos: THROUGHPUT_QOS_TIER_0
	eff darwin BG: NO
	eff iotier: THROTTLE_LEVEL_TIER0 (IMPORTANT)
	eff managed: NO
	eff qos ceiling: THREAD_QOS_UNSPECIFIED
	eff qos clamp: THREAD_QOS_UNSPECIFIED
	eff other: live-donor -> live-donor ui-is-urgent
```

When state changes, main thread is being stuck for about 1 sec and ffmpeg libavcodec/libavutil is busy processing AVPacket.
<details>
<img width="1526" alt="instruments2" src="https://user-images.githubusercontent.com/85604869/164838438-fba33f63-e20c-4109-be1a-5f9dd8cd5f75.png">
</details>

**Comment by exelix11** [comment]
Did some benchmarking and reworked thread usage, this also benefits non-mac builds as latency seems to be slightly decreased, hopefully also helps with the osx scheduler. 
When you can try [this build](https://github.com/exelix11/SysDVR/files/8547471/net6.0.zip)
Also code changes are becoming rather big so i pushed everythig on the [net6](https://github.com/exelix11/SysDVR/tree/net6) branch if you want to try doing some tweaks yourself.

**Comment by exelix11** [comment]
Btw before release need to do another test unrelated to performance and it's if latest libusbdotnet from nuget + osx libusb from brew work so i don't have to provide self built binaries for that.

**Comment by KazushiMe** [comment]
> Did some benchmarking and reworked thread usage, this also benefits non-mac builds as latency seems to be slightly decreased, hopefully also helps with the osx scheduler.

This build is working as expected, all issues resolved.

> it's if latest libusbdotnet from nuget + osx libusb from brew work so i don't have to provide self built binaries for that

libusb from brew works, but cannot get nuget running on arm64 (or easier for users to set up) as there is no mono bottle available: https://formulae.brew.sh/formula/mono#default
```
➜ brew info nuget                                            
nuget: stable 6.0.0 (bottled)
Package manager for Microsoft development platform including .NET
https://www.nuget.org/
License: MIT
==> Dependencies
Required: mono ✘
```
```
➜ brew install mono 
Error: mono: no bottle available!
You can try to install from source with:
  brew install --build-from-source mono
Please note building from source is unsupported. You will encounter build
failures with some formulae. If you experience any issues please create pull
requests instead of asking for help on Homebrew's GitHub, Twitter or any other
official channels.
```

**Comment by exelix11** [comment]
Great to hear all issues are fixed. 

Nuget is just dependency management on my end for building, what I meant is I need to make sure the official libusbdotnet on there will pick up native libusb from brew. Will post another build with that in a bit.

**Comment by exelix11** [comment]
[Here it ishttps://github.com/exelix11/SysDVR/files/8547686/net6.0.zip), this should be fairly close to the actual release. As i said now libusb is not bundled anymore so make sure it's installed with brew and that the loader paths are properly configured.

**Comment by KazushiMe** [comment]
> [Here it is](https://github.com/exelix11/SysDVR/files/8547686/net6.0.zip), this should be fairly close to the actual release. As i said now libusb is not bundled anymore so make sure it's installed with brew and that the loader paths are properly configured.

It works with libusb installed from brew.

Another question aside from SysDVR-Client: I'm wondering if the sysmodule could force console USB-C power role to UFP(Upstream Facing Port), so that it'll play nicely with double-sided USB-C cables and pull power up to 7.5W via PD protocol. For now the console would set to DFP(Downstream Facing Port) and act as a charger most of the time.

**Comment by exelix11** [comment]
I don't know, will do some googling but probably not, the API the switch os provides is fairly high level, I can only configure the descriptor and even that has limits, like I wanted to add the windows specific string so winusb would be installed automatically but it isn't possible.

**Comment by KazushiMe** [comment]
Thanks for the answer. This happens on Mac/PC laptops and Android devices so I guess it's on the console's side. Maybe the only way is to [communicate bm92t Power Delivery IC via I2C](https://github.com/CTCaer/switch-l4t-kernel-4.9/blob/linux-3.4.1-dev/drivers/misc/bm92txx.c). Have tried combining USB-C to USB-A adapter and USB-A to USB-C cable, Mac could supply 5V/1.5A via USB CDP.

**Comment by exelix11** [comment]
Configuring usb via I2C (assuming it's possible from HOS) is likely out of scope for sysdvr, if it's something libnx ever implements i'll use it but don't want to have to maintain more code that may break with hw revisions or fw updates

**Comment by exelix11** [comment]
This is now released in version 5.3
